resolver/src/name_server_pool.rs

```diff
@@ -201,151 +205,302 @@ impl ConnectionProvider for StandardConnection {
+                socket_addr: config.socket_addr,
+                timeout: options.timeout,
+                tls_dns_name: config.tls_dns_name.clone().unwrap_or_default(),
+            })),
+            #[cfg(feature = "dns-over-https")]
+            Protocol::Https => {
+                ConnectionHandleInner::Connect(Some(ConnectionHandleConnect::Https {
+                    socket_addr: config.socket_addr,
+                    timeout: options.timeout,
+                    tls_dns_name: config.tls_dns_name.clone().unwrap_or_default(),
+                }))
+            }
+            #[cfg(feature = "mdns")]
+            Protocol::Mdns => ConnectionHandleInner::Connect(Some(ConnectionHandleConnect::Mdns {
+                socket_addr: config.socket_addr,
+                timeout: options.timeout,
+            })),
+        };
+
+        ConnectionHandle(Arc::new(Mutex::new(dns_handle)))
+    }
+}
+
+#[derive(Debug)]
+enum ConnectionHandleConnect {
+    Udp {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+    },
+    Tcp {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+    },
+    #[cfg(feature = "dns-over-tls")]
+    Tls {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+        tls_dns_name: String,
+    },
+    #[cfg(feature = "dns-over-https")]
+    Https {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+        tls_dns_name: String,
+    },
+    #[cfg(feature = "mdns")]
+    Mdns {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+    },
+}
+
+impl ConnectionHandleConnect {
+    fn connect(self) -> Result<ConnectionHandleConnected, ProtoError> {
+        use self::ConnectionHandleConnect::*;
+
+        debug!("connecting: {:?}", self);
+        match self {
+            Udp {
+                socket_addr,
+                timeout,
+            } => {
+                let (stream, handle) = UdpClientStream::new(socket_addr);
                 // TODO: need config for Signer...
                 let dns_conn = DnsMultiplexer::with_timeout(
                     stream,
                     handle,
-                    options.timeout,
+                    timeout,
                     NoopMessageFinalizer::new(),
                 );
 
                 let (stream, handle) = DnsExchange::connect(dns_conn);
-                // TODO: instead of spawning here, return the stream as a "Background" type...
-                tokio::executor::spawn(stream.and_then(|stream| stream).map_err(|e| {
-                    error!("error, udp connection shutting down: {}", e);
-                }));
-
+                let stream = stream.and_then(|stream| stream).map_err(|e| {
+                    debug!("udp connection shutting down: {}", e);
+                });
                 let handle = BufDnsRequestStreamHandle::new(handle);
-                ConnectionHandle::UdpOrTcp(handle)
+
+                tokio::spawn(stream);
```

Comment:
@hawkw , a different option that might also fix this issue (but I don't know if it's more or less corrent), would be to use a handle here to the original Runtime, and thus put all of this on the some Handle to that Runtime. I think that might possibly fix the issue as well, not necessarily needing #568. But I think #568 is better, definitely simpler.?

Reply:
Regarding this idea, actually, if we leave it as is, it should make better usage of the multithreaded Runtime, which would be good.

Reply:
@bluejekyll I think that while making this code generic over a `futures::Executor` might be a good idea, it should probably be done at the same time as making _all_ places where `trust-dns-resolver` spawns futures work that way, so maybe that deserves its own issue (and perhaps some design discussion/weighing the pros and cons), and shouldn't be done in this branch?

Reply:
All of these are great suggestions... Let's file an issue to do that in the future.